### PR TITLE
Enables task log --follow on multiple tasks.

### PR DESCRIPTION
**Pull request by armandgrillet**
Tested by running this taks on a cluster: 

```
{
  "id": "/test-service",
  "backoffFactor": 1.15,
  "backoffSeconds": 1,
  "cmd": "while true; do hostname && sleep 120; done",
  "container": {
    "type": "DOCKER",
    "volumes": [],
    "docker": {
      "image": "ubuntu",
      "forcePullImage": false,
      "privileged": false,
      "parameters": []
    }
  },
  "cpus": 0.1,
  "disk": 0,
  "instances": 3,
  "mem": 128,
  "networks": [
    {
      "mode": "host"
    }
  ]
}
```

And ten running `build/darwin/dcos task log test-service --follow`.

### CODE REVIEW

File: pkg/cmd/task/task_log.go [hunk 1]
```diff
@@ -51,14 +51,50 @@ func newCmdTaskLog(ctx api.Context) *cobra.Command {
 					}
 					return logClient.PrintTask(task.TaskID.Value, file, opts)
 				}
-			}
+			} else {
+				msgChan := make(chan taskMessage)
+				errChan := make(chan error)
+
+				for _, task := range tasks {
```
Comment (rgo3) [inline pkg/cmd/task/task_log.go]: Is it possible to have the closure in a separate function? In general we might want to short this `if {...} else {...}` block by sourcing some code out to functions.

File: pkg/cmd/task/task_log.go [hunk 2]
```diff
@@ -51,14 +51,50 @@ func newCmdTaskLog(ctx api.Context) *cobra.Command {
 					}
 					return logClient.PrintTask(task.TaskID.Value, file, opts)
```
Comment (rgo3) [inline pkg/cmd/task/task_log.go]: Even though this is not part of the PR it caught my eye, how do we print multiple if we return after the first one?

**Review by rgo3 — CHANGES_REQUESTED** [review]

File: pkg/cmd/task/task_log.go (hunk 2 above)
Comment (bamarni) [reply pkg/cmd/task/task_log.go]: The follow case seems to be on the else statement below

File: pkg/cmd/task/task_log.go [hunk 3]
```diff
@@ -35,30 +35,52 @@ func newCmdTaskLog(ctx api.Context) *cobra.Command {
...
 						Format: output,
 						Skip:   -1 * lines,
 					}
 					return logClient.PrintTask(task.TaskID.Value, file, opts)
 				}
-			}
+			} else {
+				if output != "cat" {
+					output = "cat"
+					ctx.Logger().Info(`Task logs don't support output options. Defaulting to "cat"...`)
```
Comment (bamarni) [inline pkg/cmd/task/task_log.go]: This could be put before the if else statement no to duplicate with the other one.

Comment (armandgrillet) [reply pkg/cmd/task/task_log.go]: They don't have the same output limitation (first one is `short`, the second one is `cat`).

File: pkg/cmd/task/task_log.go [hunk 4]
```diff
@@ -35,30 +35,52 @@ func newCmdTaskLog(ctx api.Context) *cobra.Command {
...
+				errChan := make(chan error)
+				for _, task := range tasks {
+					go followAndlog(task.TaskID.Value, lines, file, msgChan, errChan)
+				}
+
+				lastTask := ""
+				for {
+					select {
+					case newTaskMsg := <-msgChan:
+						if string(newTaskMsg.data) != "" {
```
Comment (bamarni) [inline pkg/cmd/task/task_log.go]: As on line 106 we check for `if len(data) > 1` before sending to this channel, this check seems unnecessary.

File: pkg/cmd/task/task_log.go [hunk 5]
```diff
@@ -68,3 +88,38 @@ func newCmdTaskLog(ctx api.Context) *cobra.Command {
...
+	data string
+}
+
+type taskWriter struct {
+	task     string
+	receiver chan taskMessage
+}
+
+func (t *taskWriter) Write(data []byte) (n int, err error) {
+	// When the data is only one byte, it doesn't represent a message.
```
Comment (bamarni) [inline pkg/cmd/task/task_log.go]: what are the 1 byte data that we can receive here?

File: pkg/cmd/task/task_log.go (hunk 2 above)
Comment (bamarni) [reply pkg/cmd/task/task_log.go]: Ah, I think Robin's point is that we're on a loop here.

File: pkg/logs/client.go [hunk 6]
```diff
@@ -144,6 +123,50 @@ func (c *Client) PrintTask(taskID string, file string, opts Options) error {
...
+		return err
+	}
+	defer client.Unsubscribe(events)
+
+	for msg := range events {
+		if len(msg.Data) == 0 {
+			continue
+		}
+
+		if print {
```
Comment (bamarni) [inline pkg/logs/client.go]: ideally we don't overwrite the built-in print function

File: pkg/logs/client.go [hunk 7]
```diff
@@ -144,6 +123,50 @@ func (c *Client) PrintTask(taskID string, file string, opts Options) error {
...
+	return nil
+}
+
+func (c *Client) dumpEntry(rawEntry []byte) error {
+	var entry Entry
+	err := json.Unmarshal(rawEntry, &entry)
+	if err != nil {
+		return err
+	}
+	fmt.Fprint(c.out, entry.Fields.Message)
```
Comment (bamarni) [inline pkg/logs/client.go:167]: For the record I'm not a big fan of this approach, if someone uses the client with `os.Stdout` for example, they would get all the entries next to each other without new lines separator. There should be a well-defined API for following logs, eg. using a channel instead of an io.Writer.

File: pkg/logs/client.go [hunk 8]
```diff
@@ -211,6 +234,9 @@ func (c *Client) resetColor() {
 }
 
 func httpResponseToError(resp *http.Response) error {
+	if resp.StatusCode == 204 {
+		return fmt.Errorf("unexpected status code %d, no logs found", resp.StatusCode)
```
Comment (bamarni) [inline pkg/logs/client.go]: Could this just be `no logs found`?

**Review by bamarni — CHANGES_REQUESTED** [review]
